docs/terrain.md

```diff
@@ -0,0 +1,123 @@
+| :------------- | :-------------- | :---------------------------------------------------------------------------------- |
+| Signedness     | `[u8; 3]`       | Always the signedness of the following vectors. 0xFF in negative axes.              |
+| Offset (65536) | `[i8; 3]`       | Offset from the last chunk in the blob, each axis being multiplied by 65536 chunks. |
+| Offset (256)   | `[i8; 3]`       | Offset from the last chunk in the blob, each axis being multiplied by 256 chunks.   |
+| Offset (1)     | `[i8; 3]`       | Offset from the last chunk in the blob, each axis being in singular chunks.         |
+| Voxels         | [Voxel](#voxel) | As many voxels as needed to fulfill the maximum amount per chunk.                   |
+
+### Voxel
+
+The `Voxel` type is stored with a dynamic size (between 1-4 bytes) dependent on its set [bitflags](#voxel.flag). Only stored within [chunks](#chunk). A reference for the data contained within voxels [can be found here][Terrain.WriteVoxelChannels]. Empty voxels are stored with their material set to Air.
+
+Voxels are stored within a chunk using [run-length encoding]. A count of `1` would indicate the voxel repeating itself once, meaning 2 of the same voxel in a row would be read when decoded. Occupancy values are stored using the float format described in the [document conventions](#document-conventions).
+
+Water occupancy was added as an addendum to the existing format for the new [Shorelines] feature release. To avoid restricting the [Voxel.Flag](#voxel.flag)'s `Material Index` field to 5 bits, or rearchitecting the format, run-length encoding is disabled for the current voxel when it has a water occupancy of above `0.0`. Instead, the `Store Count` bit is enabled, and a count of `0` is written. Following the empty count, the water occupancy byte is stored. If a voxel has a solid occupancy of `1.0` (unless the material is Air), the water occupancy should always be `0.0`.
+
+| Field Name      | Format                    | Value                                                                                                          |
+| :-------------- | :------------------------ | :------------------------------------------------------------------------------------------------------------- |
+| Flag            | [Voxel.Flag](#voxel.flag) | Contains the material of this voxel, along with other bitflags.                                                |
+| Solid Occupancy | `f8`                      | Occupancy between 0-100% of the set material for this voxel. Only stored if the `Store Occupancy` flag is set. |
+| Count           | `u8`                      | Run-length count. Only stored if the `Store Count` flag is set.                                                |
+| Water Occupancy | `f8`                      | Occupancy between 0-100% of Water for this voxel. Only stored based on the conditions described above.         |
```

Comment:
I'd love an explanation of occupancy in the context of Terrain voxels, even if it's just "Occupancy represents roughly how full this voxel is as a percentage". It's better to overexplain than underexplain.